### PR TITLE
application upload - perf improvement using joblib

### PR DESCRIPTION
**Pull request by maburlik**
Fixes Bug 8657234: sfctl application upload has perf issue since it uploads files one at a time. 7.2 release validation timed out and required timeout increase for small app

Changes include:

- Updated 'application upload' native imagestore application upload to parallelize uploads, greatly reducing overall upload time for applications with many smaller files.
- Improved progress log shown by flag '--show-progress' to have progress slider, using tqdm module.

Verified:

- [x] Build and test CI passes
- [x] History and readme updated to reflect changes
- [ ] N/A Package version updated according to semantic versioning rules
- [x] Tests modified or added, when applicable
- [ ] N/A Updated code owners file, when applicable
- [x] Read the [PR checklist](https://github.com/Azure/service-fabric-cli/wiki/Checklist)

During 7.2 release validation the app uploads timed out at 5 min for some small apps, which led me to look at the implementation of application upload. Since it takes time to start file transfers for each file it's commonly beneficial to do this across multiple threads/processes.

After the change to use machine no. threads for degree of parallelism, the app that was timing out uploading at 5 min was able to be uploaded in 10 sec.

From src/test/current/scripts/linuxonebox/installgithubsamples.sh:

Uploading path: CounterServiceWebServicePkg/Code/runtimes/debian-x64: 0it [00:00, ?it/s]
[114/342] files, CounterServiceWebServicePkg/Code/runtimes/debian-x64/_.dir
Time left: 1193 seconds
Uploading path: CounterServiceWebServicePkg/Code/runtimes/debian-x64/native: 100%|█| 1/1 [00:00<00:00,  7.1
[116/342] files, CounterServiceWebServicePkg/Code/runtimes/debian-x64/native/_.dir
Time left: 1193 seconds
Uploading path: CounterServiceWebServicePkg/Code/refs: 100%|█████████████| 158/158 [00:03<00:00, 45.96it/s]
[275/342] files, CounterServiceWebServicePkg/Code/refs/_.dir
Time left: 1189 seconds
Uploading path: CounterServiceWebServicePkg/Data: 100%|██████████████████████| 1/1 [00:00<00:00,  9.33it/s]
[277/342] files, CounterServiceWebServicePkg/Data/_.dir
Time left: 1189 seconds
Uploading path: CounterServiceWebServicePkg/Config: 100%|████████████████████| 2/2 [00:00<00:00, 17.54it/s]
[280/342] files, CounterServiceWebServicePkg/Config/_.dir
Time left: 1189 seconds
Uploading path: CounterServicePkg: 100%|█████████████████████████████████████| 3/3 [00:00<00:00, 26.66it/s]
[284/342] files, CounterServicePkg/_.dir
Time left: 1189 seconds
Uploading path: CounterServicePkg/Code: 100%|████████████████████████████████| 9/9 [00:00<00:00, 43.03it/s]
[294/342] files, CounterServicePkg/Code/_.dir
Time left: 1189 seconds
Uploading path: CounterServicePkg/Code/runtimes: 0it [00:00, ?it/s]
[295/342] files, CounterServicePkg/Code/runtimes/_.dir
Time left: 1188 seconds
Uploading path: CounterServicePkg/Code/runtimes/linux: 0it [00:00, ?it/s]
[296/342] files, CounterServicePkg/Code/runtimes/linux/_.dir
Time left: 1188 seconds
Uploading path: CounterServicePkg/Code/runtimes/linux/lib: 0it [00:00, ?it/s]
[297/342] files, CounterServicePkg/Code/runtimes/linux/lib/_.dir
Time left: 1188 seconds
Uploading path: CounterServicePkg/Code/runtimes/linux/lib/netstandard2.0: 100%|█| 21/21 [00:01<00:00, 17.30
[319/342] files, CounterServicePkg/Code/runtimes/linux/lib/netstandard2.0/_.dir
Time left: 1187 seconds
Uploading path: CounterServicePkg/Code/runtimes/win: 0it [00:00, ?it/s]
[320/342] files, CounterServicePkg/Code/runtimes/win/_.dir
Time left: 1187 seconds
Uploading path: CounterServicePkg/Code/runtimes/win/lib: 0it [00:00, ?it/s]
[321/342] files, CounterServicePkg/Code/runtimes/win/lib/_.dir
Time left: 1187 seconds
Uploading path: CounterServicePkg/Code/runtimes/win/lib/netstandard2.0: 100%|█| 15/15 [00:00<00:00, 35.07it
[337/342] files, CounterServicePkg/Code/runtimes/win/lib/netstandard2.0/_.dir
Time left: 1186 seconds
Uploading path: CounterServicePkg/Data: 100%|████████████████████████████████| 1/1 [00:00<00:00,  5.69it/s]
[339/342] files, CounterServicePkg/Data/_.dir
Time left: 1186 seconds
Uploading path: CounterServicePkg/Config: 100%|██████████████████████████████| 2/2 [00:00<00:00, 18.08it/s]
[342/342] files, CounterServicePkg/Config/_.dir
Time left: 1186 seconds
Complete

### CODE REVIEW

**Comment by jeffj6123** [comment]
@samedder where can we check to make sure these 2 new libraries are allowed to be used?

File: requirements.txt [hunk 1]
```diff
@@ -5,4 +5,6 @@ pylint
 vcrpy
 contextlib2
 mock
-jsonpickle
\ No newline at end of file
+jsonpickle
```
Comment (jeffj6123) [inline requirements.txt:8]: Remove the jsonpickle requirement. This shouldnt be needed anymore as it was removed previously and we are not allowed to use anything besides the standard json lib.

File: src/sfctl/custom_app.py [hunk 2]
```diff
@@ -50,6 +77,15 @@ def path_from_imagestore_string(imagestore_connstr):
         return conn_str_list[1]
     return False
 
+def get_job_count():
+    """
+    Test-mockable wrapper for returning cpu count.
+    """
+    jobcount = cpu_count()
```
Comment (samedder) [inline src/sfctl/custom_app.py:84]: I believe this can throw also, should put it in a try I believe.

File: src/sfctl/custom_app.py [hunk 3]
```diff
@@ -101,6 +137,45 @@ def get_lesser(num_a, num_b):
...
+                'timeout': current_time_left})
+        url = urlunparse(url_parsed)
+
+        # timeout is (connect_timeout, read_timeout)
+        res = sesh.put(url, data=file_opened,
+                        timeout=(get_lesser(60, current_time_left), current_time_left))
+
+        res.raise_for_status()
+
 def upload_to_native_imagestore(sesh, endpoint, abspath, basename, #pylint: disable=too-many-locals,too-many-arguments
```
Comment (samedder) [inline src/sfctl/custom_app.py:179]: I would say also make the number of parallel uploads an optional argument

Comment (samedder) [inline src/sfctl/custom_app.py:179]: Please also add help description changes for the command

**Review by samedder — DISMISSED** [review]

**Comment by samedder** [comment]
> @samedder where can we check to make sure these 2 new libraries are allowed to be used?

There's two options, you can merge it and see if internally it generates a bug for insecure package use. That will require a PR to follow up. Or, you can do what I did and try to find another repo already using them and verify they don't have an open issue.

I looked manually now though and it seems the packages are O.K. but I didn't check the specific version.

File: src/sfctl/custom_app.py [hunk 4]
```diff
@@ -10,16 +10,43 @@
...
+from multiprocessing import Process, cpu_count
 from time import time
 import sys
 import zipfile
 import shutil
 import xml.etree.ElementTree as ET
+import contextlib
 from knack.util import CLIError
+import joblib
+from joblib import Parallel, delayed
```
[the remark this answers is not in the document]
Comment (maburlik) [reply src/sfctl/custom_app.py:22]: Need import joblib due to the tqdm callback logic, and the 'from' line was to make the Parallel, delayed calls more concise. I would need to either add extra things to the from or lengthen the explicit calls for componenents under joblib. What do you prefer?

File: src/sfctl/custom_app.py (hunk 2 above)
Comment (maburlik) [reply src/sfctl/custom_app.py:84]: ok

File: src/sfctl/custom_app.py (hunk 3 above)
Comment (maburlik) [reply src/sfctl/custom_app.py:179]: Is there a practical scenario a customer would want this? For each file transfer there is an initial round trip time & file transfer turnaround period of exponential rate ramp up (to throughput cap), which results in slow transfer today when an app has even a relatively small number of files (<100).

Comment (maburlik) [reply src/sfctl/custom_app.py:179]: Will do. You mean just for the parameters?

**Review by jeffj6123 — APPROVED** [review]

**Review by iyyappam — APPROVED** [review]
:shipit:

**Pull request merged**